internal/core/processor/universal/processor.go

```diff
@@ -193,3 +304,32 @@ func getLastBlockData(ctx context.Context, client blockchain.EthClient, lastBloc
 		Hash:      header.Hash(),
 	}, nil
 }
+
+func checkout(tx state.Tx, contractAddress common.Address, blockNumber uint64) error {
+	err := tx.LoadMerkleTrees(contractAddress)
+	if err != nil {
+		return err
+	}
+
+	err = tx.Checkout(contractAddress, int64(blockNumber))
+	if err != nil {
+		slog.Error("error occurred checking out merkle tree at block number", "block_number", blockNumber,
+			"contract_address", contractAddress, "err", err)
+		return err
+	}
+
+	return nil
+}
+
+func getSafeBlock(currentBlockNumber uint64) *model.Block {
```

Comment:
I think this function could be simplified like this

```go
	if currentBlockNumber < safeBlockMargin {
		return &model.Block{ Number: 0}

	return &model.Block{ Number: currentBlockNumber - safeBlockMargin}
```

Reply:
can you rechecked this. as it is implemented now I think you have a bug. 

previously you were returning 0 in case that currentBlock is less then blockMargin. and now you return currentBlockNumber.